contracts/golos.vesting/golos.vesting.cpp

```diff
@@ -0,0 +1,500 @@
+#include "golos.vesting.hpp"
+#include <eosiolib/transaction.hpp>
+#include <eosiolib/fixedpoint.hpp>
+
+using namespace eosio;
+
+extern "C" {
+    void apply(uint64_t receiver, uint64_t code, uint64_t action) {
+        vesting(receiver).apply(code, action);
+    }
+}
+
+void vesting::apply(uint64_t code, uint64_t action) {
+    if (N(transfer) == action && N(eosio.token) == code)
+        execute_action( this, &vesting::buy_vesting );
+
+    else if (N(accruevg) == action)
+        execute_action(this, &vesting::accrue_vesting);
+    else if (N(convertvg) == action)
+        execute_action(this, &vesting::convert_vesting);
+    else if (N(delegatevg) == action)
+        execute_action(this, &vesting::delegate_vesting);
+    else if (N(undelegatevg) == action)
+        execute_action(this, &vesting::undelegate_vesting);
+    else if (N(cancelvg) == action)
+        execute_action(this, &vesting::cancel_convert_vesting);
+    else if (N(expbattery) == action)
+        execute_action(this, &vesting::expend_battery);
+
+    else if (N(issue) == action)
+//        issue(unpack_action_data<structures::issue_vesting>());
+        execute_action(this, &vesting::issue);
+    else if (N(createpair) == action)
+        execute_action(this, &vesting::create_pair);
+    else if (N(timeoutrdel) == action)
+        calculate_delegate_vesting();
+    else if (N(timeoutconv) == action)
+        calculate_convert_vesting();
+    else if (N(timeout) == action)
+        timeout_delay_trx();
+}
+
+void vesting::buy_vesting(account_name  from,
+                          account_name  to,
+                          asset         quantity,
+                          string        /*memo*/) {
+    if(_self != from) // TODO account golos.vesting can not buy vesting
+        require_auth(from);
+    else return;
+
+    auto sym_name = quantity.symbol.name();
+    tables::vesting_table table_pair(_self, _self);
+    auto index = table_pair.get_index<N(token)>();
+    auto pair = index.find(sym_name);
+    eosio_assert(pair != index.end(), "Token not found");
+
+    issue({_self, convert_token(quantity, pair->vesting.symbol)}, false, true);
+    transfer(to, from, quantity, false);
```

Comment:
As I can see `issue` and `transfer` using only in this place. May be it is useful to implement it actions in place? Especially the use of boolean flags makes a lot of confusion.